qiskit/dagcircuit/dagnode.py

```diff
@@ -259,9 +260,13 @@ def __init__(self, op, qargs: Iterable[Qubit] = (), cargs: Iterable[Clbit] = (),
         self.qargs = tuple(qargs)
         self.cargs = tuple(cargs)
         if dag is not None:
-            self.sort_key = ",".join(
-                f"{dag.find_bit(q).index:04d}" for q in itertools.chain(self.qargs, self.cargs)
-            )
+            qargs = tuple(itertools.chain(self.qargs, self.cargs))
+            key = self._key_cache.get(qargs, None)
+            if key is not None:
+                self.sort_key = key
+            else:
+                self.sort_key = ",".join(f"{dag.find_bit(q).index:04d}" for q in qargs)
+                self._key_cache[qargs] = self.sort_key
```

Comment:
I don't think this cache lookup works; the same qubit instances could have different indices between two DAGs, but all the caching is shared per interpreter.  Switching to having this be an instance-level cache on the DAG would solve that, and let them be garbage collected as well.

Reply:
That works for me, I put a paragraph in the commit message about how I didn't think the global cache really mattered as long as the sort was consistent within a dag. But since we get the dag in `__init__` anyway we can just move it there and side step that potential issue.

Reply:
Also, I think we might want to see if we can work out a scheme where we do the key lookup based on the `qargs` and `cargs` tuples being shared separately - I know certainly for strings, the hash lookup is faster if you're using the referentially identical object to do the lookup, and we're moving towards a place where the `qargs` tuples _will_ be shared, and the `cargs` ones most often are already (since they're the empty tuple, which is a singleton).  I don't know for sure if tuples have the same behaviour, but at any rate, the `tuple(itertools.chain(...))` call is annoyingly slow if we can organise our cache to avoid it.

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/10729/commits/c57207d38a1e4fd38dc52829c20a920349903b46